### PR TITLE
Add set operations for `NameSet` and `TypeSet`.

### PR DESCRIPTION
**Pull request by jmalloc**
This PR adds various set operations for `NameSet` and `TypeSet`.  Where possible the implementations work generically on the `NameCollection` and `TypeCollection` interfaces.

### CODE REVIEW

File: message/typeset.go [hunk 1]
```diff
@@ -30,6 +30,65 @@ func TypesOf(messages ...dogma.Message) TypeSet {
 	return s
 }
 
+// IntersectionT returns the intersection of the given collections.
+//
+// That is, it returns a set containing only those types that are present in all
```
Comment (danilvpetrov) [inline message/typeset.go]: Should it mention that if the given collections are not intersecting, then an empty set is returned?

File: message/typeset.go [hunk 2]
```diff
@@ -30,6 +30,65 @@ func TypesOf(messages ...dogma.Message) TypeSet {
...
+		})
+	}
+
+	return r
+}
+
+// DiffT returns the difference between a and b.
+//
+// That is, it returns a set containing all of the types present in a that are
+// not present in b.
```
Comment (danilvpetrov) [inline message/typeset.go]: Should it mention that if `b` does not have any differences to `a`, then an empty set is returned?

File: message/typeset.go (hunk 1 above)
Comment (jmalloc) [reply message/typeset.go]: I didn't think this needed to be explained separately as it's not a special case.

File: message/typeset.go (hunk 2 above)
Comment (jmalloc) [reply message/typeset.go]: Again it's not a special case -- that's also true if `b` is a proper superset of `a`.  Maybe I should link to some external docs on the various set operations?

I am re-reading these docs though, and I think I should at least explain what happens when you don't pass any collections at all.

**Pull request merged**